Add Vitals Estimates page (#3544)
Kong/docs.konghq.com: app/gateway/2.7.x/vitals/vitals-estimates.md

---
title: Estimate Vitals Storage in PostgreSQL
badge: enterprise
---

Vitals data can be divided into two categories: Kong Gateway node statistics and request response codes.

## Kong Gateway node statistics

> These types of metrics are proxy latency, upstream latency, and cache hit/miss. Kong Gateway node statistics are stored in tables like the following:

Reviewer: ```suggestion
These types of metrics are proxy latency, upstream latency, and cache hit/miss. {{site.base_gateway}} node statistics are stored in tables like the following:
```
Can you replace any instances of "Kong Gateway" with the variable above?

> * `vitals_stats_seconds_timestamp` stores 1 new row for every _second_ Kong runs

Reviewer: ```suggestion
* `vitals_stats_seconds_timestamp` stores 1 new row for every _second_ Kong runs.
```
Or, no period in the other two bullet points in this list. For consistency, maybe no periods, as the other lists on the page don't have them (eg lines 20-22)?

* `vitals_stats_minutes` stores 1 new row for every _minute_ Kong runs.
* `vitals_stats_days` stores 1 new row for every _day_ Kong runs.

Kong Gateway node statistics are not associated with specific Kong Gateway entities like Workspaces, Services, or Routes. They're designed to represent the cluster's state in time. This means the tables will have new rows regardless if Kong Gateway is routing traffic or idle.

The tables do not grow infinitely and hold data for the following duration of time:

* `vitals_stats_seconds_timestamp` holds data for 1 hour (3600 rows)
* `vitals_stats_minutes` holds data for 25 hours (90000 rows)
* `vitals_stats_days` holds data for 2 years (730 rows)

## Request response codes

Request response codes are stored in the other group of tables following a different rationale. Tables in this group share the same structure (`entity_id`, `at`, `duration`, `status_code`, `count`):

* `vitals_code_classes_by_workspace`
* `vitals_code_classes_by_cluster`
* `vitals_codes_by_route`

> The `entity_id` does not exist in `vitals_code_classes_by_cluster` as it doesn't store per any entity.

Reviewer: "doesn't store per any entity" is confusing to me. I'm guessing that it's meant to say something like "this table doesn't store entity-specific information".

> `entity_id` is `workspace_id` in `vitals_code_classes_by_workspace` and `service_id`, `route_id` in `vitals_codes_by_route`.

Reviewer: I'd turn this phrasing around:
```suggestion
In the `vitals_code_classes_by_workspace` table, `entity_id` is `workspace_id`. In the `vitals_codes_by_route` table, `entity_id` is `service_id` and `route_id`. 
```

> `at`, a timestamp, is the start of the period a row represents, while `duration` is the duration of that period.

Reviewer: ```suggestion
`at` is a timestamp. It logs the start of the period a row represents, while `duration` is the duration of that period.
```

`status_code` and `count` are the quantity of the HTTP status codes (200, 300, 400, 500) observed in the period represented by a row.

> While Kong Gateway node statistic tables grow only according to time, status code tables will only have new rows when Kong Gateway proxies traffic, and the number of new rows depends on the traffic itself.

Reviewer: ```suggestion
While Kong Gateway node statistic tables grow only according to time, status code tables only have new rows when Kong Gateway proxies traffic, and the number of new rows depends on the traffic itself.
```

## Example

Consider a brand new Kong Gateway that hasn't proxied any traffic yet. Kong Gateway node statistic tables have rows but status codes tables don't.

When Kong Gateway proxies its first request at `t` returning status code 200, the following rows are added:

```bash
vitals_codes_by_cluster
[second(t), 1, 200, 1]
[minute(t), 60, 200, 1] 
[day(t), 84600, 200, 1]
```

> Second, minute, and day content is trimmed in the follow way:

Reviewer: ```suggestion
Second, minute, and day content is trimmed in the following way:
```

```bash
vitals_codes_by_cluster
[second(t), 1, 200, 1]
[second(t + 5s), 1, 200, 1]
[minute(t), 60, 200, 2]
[day(t), 84600, 200, 2]

vitals_codes_by_workspace
[workspace_id, second(t), 1, 200, 1]
[workspace_id, second(t + 5s), 1, 200, 1]
[workspace_id, minute(t), 60, 200, 2]
[workspace_id, day(t), 84600, 200, 2]

vitals_codes_by_route
[service_id, route_id, second(t), 1, 200, 1]
[service_id, route_id, second(t + 5s), 1, 200, 1]
[service_id, route_id, minute(t), 60, 200, 2]
[service_id, route_id, day(t), 84600, 200, 2]
```

In summary, the number of rows in those status codes tables is directly related to:

* The number of status codes observed in Kong Gateway proxied requests
* The number of Kong Gateway entities involved in those requests
* The constant flow of proxied requests

In an estimate of row numbers in scenario, consider a Kong Gateway cluster with the following characteristics:  

* A constant flow of requests returning all 5 possible groups of status codes (1xx, 2xx, 3xx, 4xx and 5xx).
* Just 1 workspace, 1 service, and 1 route

After 24 hours of traffic, the status codes tables will have this number of rows:

> * **vitals_codes_by_cluster**: 5 row for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.

Reviewer: ```suggestion
* **vitals_codes_by_cluster**: 5 rows for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.
```

> * **vitals_codes_by_cluster**: 5 row for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.

Reviewer: This would be better presented in a table, with the headings: status code table | day | minute | seconds | total

> * **vitals_codes_by_workspace**: 5 row for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.

Reviewer: ```suggestion
* **vitals_codes_by_workspace**: 5 rows for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.
```

> * **vitals_codes_by_route**: 5 row for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.

Reviewer: ```suggestion
* **vitals_codes_by_route**: 5 rows for day, (5 * 24 * 60 = 7200) rows for minutes and (5 * 60 * 60 = 18000) rows for seconds. Total = 25200 rows.
```
